experimental/form_app/lib/src/form_widgets.dart

```diff
@@ -0,0 +1,208 @@
+import 'package:flutter/material.dart';
+import 'package:intl/intl.dart' as intl;
+
+class FormWidgetsDemo extends StatefulWidget {
+  @override
+  _FormWidgetsDemoState createState() => _FormWidgetsDemoState();
+}
+
+class _FormWidgetsDemoState extends State<FormWidgetsDemo> {
+  final _formKey = GlobalKey<FormState>();
+  String title = "";
+  String description = "";
+  DateTime date;
+  double maxValue = 0;
+  bool brushedTeeth = false;
+  bool enableFeature = false;
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        title: Text('Form widgets'),
+      ),
+      body: Form(
+        key: _formKey,
+        child: Scrollbar(
+          child: Align(
+            alignment: Alignment.topCenter,
+            child: Card(
+              child: SingleChildScrollView(
+                padding: EdgeInsets.all(16),
+                child: ConstrainedBox(
+                  constraints: BoxConstraints(maxWidth: 400),
+                  child: Column(
+                    mainAxisAlignment: MainAxisAlignment.center,
+                    crossAxisAlignment: CrossAxisAlignment.center,
+                    children: [
+                      ...[
+                        TextFormField(
+                          decoration: InputDecoration(
+                            filled: true,
+                            hintText: 'Enter a title...',
+                            labelText: 'Title',
+                          ),
+                          onChanged: (value) {
+                            title = value;
+                          },
+                        ),
+                        TextFormField(
+                          decoration: InputDecoration(
+                            border: const OutlineInputBorder(),
+                            filled: true,
+                            hintText: 'Enter a description...',
+                            labelText: 'Description',
+                          ),
+                          onChanged: (value) {
+                            description = value;
+                          },
+                          maxLines: 5,
+                        ),
+                        _FormDatePicker(
+                          onChanged: (value) {
+                            date = value;
```

Comment:
This is modifying state without calling setState, which is a problem. Widgets like this are weird, because they have complex UI state to represent the value, but need to still report the data state to their parent.

The pattern used in the framework for a lot of widgets like this is to have the parent widget keep track of the state and then inject it back into the child when it's changed. The Slider that's a few lines down does this:

```dart
Slider(
  min: 0,
  max: 500,
  divisions: 500,
  value: maxValue,
  onChanged: (value) {
    setState(() {
      maxValue = value;
    });
  },
),
```

When the user does a gesture to move the slider, there's a little dance that's done:

* Slider calls its onChanged method.
* The parent updates its field for the slider value in setState (`maxValue` in this case).
* The parent's build method is called, creating a new Slider with the new value.
* Slider rebuilds to display the new data, which causes the little knob to actually move onscreen.

I believe you could put the same flow to work here.

Reply:
That makes more sense.  d0aaf7add5133be9c753f4286bcf9789272abcba